### DISCUSSION TITLE
Disable IdeaVim in terminals and consoles

### DISCUSSION THREAD

**Discussion by bruno-robert** [JetBrains/ideavim#452] [Q&A]
Is there a way to disable IdeaVim in terminals (or to only enable it for editors)?

**Comment by geoffreyvanwyk** [comment]
This is crucial, because if it cannot be disabled, then it is impossible to to use vi-mode for command line editing and also impossible to use vim in the integrated terminal.

If it cannot be disabled, the integrated terminal becomes a vestigial component of the IDE and consequently a waste of hard-earned money spent on the subscription.

**Comment by geoffreyvanwyk** [comment]
I found the workaround: https://youtrack.jetbrains.com/issue/VIM-2670/Vim-mode-in-IDE-terminal-shell#focus=Comments-27-6160010.0-0 Change the key binding for switching from Terminal to Editor from ESC to something else or remove it.

**Reply by huyz** [reply]
@bruno-robert  This is the answer.  It's not a problem with IdeaVim. it's a problem with the default "Switch Focus to Editor" keybinding which is set to `ESC` ⎋.

So the problem is a bad IntelliJ default and the solution is a simple rebinding:
![screenshot 2023-11-24T102518Z](https://github.com/JetBrains/ideavim/assets/128394/49350012-7955-4485-ac9b-691e8a44ebc8)

**Comment by wesleylether** [comment]
I still have issues with vim in the console or terminal.
It would be great if you can configure this. The switch focus to editor keybinding change is a hack not a solution.

It would be great If we just can disable VIM for console and terminal!

**Reply by citizenmatt** [reply]
Changing the shortcut to return focus from the Terminal tool window from Escape is not a hack - it's the only way to allow the Terminal window to process Escape itself, and allow the Terminal's shell to use vi mode. This is built in to the platform and not related to IdeaVim.

Also, IdeaVim is not normally active in the Terminal tool window. Is your request related to something else, such as the new Terminal, which uses IDE based popup completion?

**Reply by wesleylether** [reply]
Ahh, yes! I was creating a new ticket. But I saw this ticket.
But for a new project, I must use Python and so PyCharm Professional with IdeaVim.

However, when I run a script, the run console also has Vim enabled. However, this doesn't work good with vim enabled. So It would be perfect if the run / debug console doesn't use vim and the editor itself does use it.

Also, in debug mode, there is auto-complete enabled. So when you type something, the auto-complete suggestions window is visible. However, when you hit escape, the window hides but also insert mode.

Is it a big feature to make it configurable where vim mode is enabled and disabled?

**Reply by citizenmatt** [reply]
Thanks for the use case. It would be helpful to hear why IdeaVim doesn't work so well in the run/debug stdin/stdout consoles.

I'm not sure that there's a reliable way to know if an editor is being used as stdin/stdout. IdeaVim tries to recognise editors that are. being used for other reasons, such as tailing a log file or interactive stdin/stdout. If these additional editors are writable, it should automatically enter Insert mode. Does this not work for your Python scripts?

The [`'ideavimsupport'` option](https://github.com/JetBrains/ideavim/wiki/set-commands#ideavim-only-options) allows enabling/disabling Vim support in different types of editors, but currently only editors inside a 'dialog', or 'singleline' editors. I wonder if it's possible to make these checks specific enough to add something like `console`, which would be the stdin/stdout of runconfigurations? 🤔

**Reply by wesleylether** [reply]
I do validate that it automatically enters insert mode. However, in the debug console, when you type something, the auto-suggestions popup will appear. But when you close that popup with ESC, you also return to ‘normal’ mode.

I never use VIM in consoles or terminals, so it would be amazing if this were configurable!

I understand that this might be a challenging task to engineer, but I hope it will be possible in the future. 😎

**Comment by anton-daneyko-ultramarin** [comment]
I have a similar issue and what could help me is a shortcut to toggle "Enable" in the settings, see screenshot:
<img width="716" alt="image" src="https://github.com/user-attachments/assets/d8bfaf17-82a8-4748-b80c-d5f18220d513" />

**Reply by citizenmatt** [reply]
Can you confirm the problem you're seeing? The latest version of IdeaVim (2.19.0) will automatically disable Vim behaviour in the new editor-based Terminal view. If you're seeing other problems, please let us know.

But you can always assign a shortcut to the "Vim" IDE action, and this will allow toggling Vim functionality for all editors. 

<img width="982" alt="Screenshot 2025-02-13 at 9 14 27 am" src="https://github.com/user-attachments/assets/75cab69e-604b-4353-993f-e7f20a6be2d2" />

**Reply by wesleylether** [reply]
Hi @citizenmatt ,

In the terminal view it works like a charm. But the problem still occurs with the consoles (* also in debug).
So when you run or debug a python script in pycharm the vim is also enabled in the consoles. Those places you also want to disable vim. Or have the ability to disable them :D

**Reply by citizenmatt** [reply]
Thanks @wesleylether. Have replied to this point in another thread

**Comment by mariolopjr** [comment]
Hi! I also have a similar issue to this. While I don't have VIM-like motions in the terminal (I believe I have the 'new' terminal experience enabled), the `which-key` plugin activates when the terminal is active. I have `<leader>` mapped to space so it makes the terminal difficult to use lol

**Comment by Thogor** [comment]
IdeaVim seems to be recently disabled in consoles and terminals now, but I actually want it to be enabled in consoles for easy searching and yanking stuff from it. 

Would it be possible to expand the 'ideavimsupport' option with "terminal" and "console" so people can choose whether or not they want IdeaVim active in these contexts?

**Reply by lzb487558900** [reply]
Hello, could you please tell me if you have found the way to open vim from the console?

**Reply by msp98** [reply]
Yes, an option would be great! Please let me know if you find a way to enable it somehow

**Comment by jgjl** [comment]
I just learned that IdeaVim is supposed to be disabled in consoles and terminal since version 2.19. I'm using IdeaVim Version 2.30 right now, and it insists to be enabled in my R console. Does the specific console plugin have to do something to get the behavior of IdeaVim disabling itself?
